Шаг 1: Загрузка и первичный анализ данных, подготовка данных

In [1]:
import pandas as pd

from sklearn.utils import shuffle
from sklearn.model_selection import train_test_split
from sklearn.metrics import accuracy_score
from sklearn.metrics import f1_score
from sklearn.linear_model import LogisticRegression
from sklearn.ensemble import RandomForestClassifier
from sklearn.preprocessing import MinMaxScaler
from sklearn.metrics import roc_auc_score
from sklearn.metrics import roc_curve 

import seaborn as sns
import matplotlib.pyplot as plt

In [2]:
data = pd.read_csv('Churn.csv')

In [3]:
data.head(15)

RowNumber  CustomerId   Surname  CreditScore Geography  Gender  Age   
0           1    15634602  Hargrave          619    France  Female   42  \
1           2    15647311      Hill          608     Spain  Female   41   
2           3    15619304      Onio          502    France  Female   42   
3           4    15701354      Boni          699    France  Female   39   
4           5    15737888  Mitchell          850     Spain  Female   43   
5           6    15574012       Chu          645     Spain    Male   44   
6           7    15592531  Bartlett          822    France    Male   50   
7           8    15656148    Obinna          376   Germany  Female   29   
8           9    15792365        He          501    France    Male   44   
9          10    15592389        H?          684    France    Male   27   
10         11    15767821    Bearce          528    France    Male   31   
11         12    15737173   Andrews          497     Spain    Male   24   
12         13    15632264       Kay          476    France  Female   34   
13         14    15691483      Chin          549    France  Female   25   
14         15    15600882     Scott          635     Spain  Female   35   

    Tenure    Balance  NumOfProducts  HasCrCard  IsActiveMember   
0      2.0       0.00              1          1               1  \
1      1.0   83807.86              1          0               1   
2      8.0  159660.80              3          1               0   
3      1.0       0.00              2          0               0   
4      2.0  125510.82              1          1               1   
5      8.0  113755.78              2          1               0   
6      7.0       0.00              2          1               1   
7      4.0  115046.74              4          1               0   
8      4.0  142051.07              2          0               1   
9      2.0  134603.88              1          1               1   
10     6.0  102016.72              2          0               0   
11     3.0       0.00              2          1               0   
12    10.0       0.00              2          1               0   
13     5.0       0.00              2          0               0   
14     7.0       0.00              2          1               1   

    EstimatedSalary  Exited  
0         101348.88       1  
1         112542.58       0  
2         113931.57       1  
3          93826.63       0  
4          79084.10       0  
5         149756.71       1  
6          10062.80       0  
7         119346.88       1  
8          74940.50       0  
9          71725.73       0  
10         80181.12       0  
11         76390.01       0  
12         26260.98       0  
13        190857.79       0  
14         65951.65       0

In [4]:
data.describe()

RowNumber    CustomerId   CreditScore           Age       Tenure   
count  10000.00000  1.000000e+04  10000.000000  10000.000000  9091.000000  \
mean    5000.50000  1.569094e+07    650.528800     38.921800     4.997690   
std     2886.89568  7.193619e+04     96.653299     10.487806     2.894723   
min        1.00000  1.556570e+07    350.000000     18.000000     0.000000   
25%     2500.75000  1.562853e+07    584.000000     32.000000     2.000000   
50%     5000.50000  1.569074e+07    652.000000     37.000000     5.000000   
75%     7500.25000  1.575323e+07    718.000000     44.000000     7.000000   
max    10000.00000  1.581569e+07    850.000000     92.000000    10.000000   

             Balance  NumOfProducts    HasCrCard  IsActiveMember   
count   10000.000000   10000.000000  10000.00000    10000.000000  \
mean    76485.889288       1.530200      0.70550        0.515100   
std     62397.405202       0.581654      0.45584        0.499797   
min         0.000000       1.000000      0.00000        0.000000   
25%         0.000000       1.000000      0.00000        0.000000   
50%     97198.540000       1.000000      1.00000        1.000000   
75%    127644.240000       2.000000      1.00000        1.000000   
max    250898.090000       4.000000      1.00000        1.000000   

       EstimatedSalary        Exited  
count     10000.000000  10000.000000  
mean     100090.239881      0.203700  
std       57510.492818      0.402769  
min          11.580000      0.000000  
25%       51002.110000      0.000000  
50%      100193.915000      0.000000  
75%      149388.247500      0.000000  
max      199992.480000      1.000000

In [5]:
data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 10000 entries, 0 to 9999
Data columns (total 14 columns):
 #   Column           Non-Null Count  Dtype  
---  ------           --------------  -----  
 0   RowNumber        10000 non-null  int64  
 1   CustomerId       10000 non-null  int64  
 2   Surname          10000 non-null  object 
 3   CreditScore      10000 non-null  int64  
 4   Geography        10000 non-null  object 
 5   Gender           10000 non-null  object 
 6   Age              10000 non-null  int64  
 7   Tenure           9091 non-null   float64
 8   Balance          10000 non-null  float64
 9   NumOfProducts    10000 non-null  int64  
 10  HasCrCard        10000 non-null  int64  
 11  IsActiveMember   10000 non-null  int64  
 12  EstimatedSalary  10000 non-null  float64
 13  Exited           10000 non-null  int64  
dtypes: float64(3), int64(8), object(3)
memory usage: 1.1+ MB


In [6]:
data['Tenure']

0        2.0
1        1.0
2        8.0
3        1.0
4        2.0
        ... 
9995     5.0
9996    10.0
9997     7.0
9998     3.0
9999     NaN
Name: Tenure, Length: 10000, dtype: float64

In [7]:
data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 10000 entries, 0 to 9999
Data columns (total 14 columns):
 #   Column           Non-Null Count  Dtype  
---  ------           --------------  -----  
 0   RowNumber        10000 non-null  int64  
 1   CustomerId       10000 non-null  int64  
 2   Surname          10000 non-null  object 
 3   CreditScore      10000 non-null  int64  
 4   Geography        10000 non-null  object 
 5   Gender           10000 non-null  object 
 6   Age              10000 non-null  int64  
 7   Tenure           9091 non-null   float64
 8   Balance          10000 non-null  float64
 9   NumOfProducts    10000 non-null  int64  
 10  HasCrCard        10000 non-null  int64  
 11  IsActiveMember   10000 non-null  int64  
 12  EstimatedSalary  10000 non-null  float64
 13  Exited           10000 non-null  int64  
dtypes: float64(3), int64(8), object(3)
memory usage: 1.1+ MB


In [8]:
data.duplicated().sum()

0

Явных дубликатов нет.

In [9]:
data.isna().sum()

RowNumber            0
CustomerId           0
Surname              0
CreditScore          0
Geography            0
Gender               0
Age                  0
Tenure             909
Balance              0
NumOfProducts        0
HasCrCard            0
IsActiveMember       0
EstimatedSalary      0
Exited               0
dtype: int64

In [10]:
data['IsActiveMember'].value_counts()

IsActiveMember
1    5151
0    4849
Name: count, dtype: int64

In [11]:
data.drop(['RowNumber','CustomerId','Surname'], axis=1, inplace=True)

Удалим признаки, которые никак не влияют.

In [20]:
#sns.heatmap(data.corr(), annot=True, fmt='.2f', cmap='RdYlGn');

In [16]:
data['Gender'].value_counts()

Gender
Male      5457
Female    4543
Name: count, dtype: int64

In [17]:
data['Geography'].value_counts()

Geography
France     5014
Germany    2509
Spain      2477
Name: count, dtype: int64

In [18]:
data['Exited'].value_counts()

Exited
0    7963
1    2037
Name: count, dtype: int64

Ушло пока не такое страшное число людей, так что пока еще не все страшно.

Большой дисбаланс по отношению ушедших в отрицательный класс, только в гендере плюс минус одинаково. Классы нужно будет балансировать.

## Шаг 2:

In [21]:
def check_ratio(check_feature):
    a = data.groupby([check_feature, 'Exited'])['Balance'].count().rename('ratio')
    b = a /  a.groupby(level=0).sum() 
    b = pd.DataFrame(data=b)
    return b

In [22]:
check_ratio('Geography')

ratio
Geography Exited          
France    0       0.838452
          1       0.161548
Germany   0       0.675568
          1       0.324432
Spain     0       0.833266
          1       0.166734

Немцы сильно чаще расстаргают договорЮ нежели жители других стран

In [23]:
check_ratio('Gender')

ratio
Gender Exited          
Female 0       0.749285
       1       0.250715
Male   0       0.835441
       1       0.164559

Женщины немного чаще расторгают договор чем мужчины

In [24]:
check_ratio('NumOfProducts')

ratio
NumOfProducts Exited          
1             0       0.722856
              1       0.277144
2             0       0.924183
              1       0.075817
3             0       0.172932
              1       0.827068
4             1       1.000000

Нет ни одного клиента с 4 продуктами, кто не расторг договор, странно, и повышение количество продуктов тоже не свидельствует о том что человек останется.

In [26]:
check_ratio('HasCrCard')

ratio
HasCrCard Exited          
0         0       0.791851
          1       0.208149
1         0       0.798157
          1       0.201843

Тут ничего аномального, все в балансе

In [27]:
check_ratio('Tenure')

ratio
Tenure Exited          
0.0    0       0.764398
       1       0.235602
1.0    0       0.776261
       1       0.223739
2.0    0       0.805263
       1       0.194737
3.0    0       0.786638
       1       0.213362
4.0    0       0.792090
       1       0.207910
5.0    0       0.798274
       1       0.201726
6.0    0       0.799092
       1       0.200908
7.0    0       0.827027
       1       0.172973
8.0    0       0.813505
       1       0.186495
9.0    0       0.780045
       1       0.219955
10.0   0       0.793722
       1       0.206278

Вне зависимости от длительности обслуживания, клиент может расторгнуть договор прмиерно в одинаковых процентов случаев.

In [29]:
check_ratio('IsActiveMember')

ratio
IsActiveMember Exited          
0              0       0.731491
               1       0.268509
1              0       0.857309
               1       0.142691

Тут есть небоьшой дисбаланс, но это, думаю, не критично.

In [30]:
gender_one_hot = pd.get_dummies(data['Gender'], drop_first=True)
country_one_hot = pd.get_dummies(data['Geography'], drop_first=True)
tenure_one_hot = pd.get_dummies(data['Tenure'], drop_first=True)
data.drop(['Gender', 'Geography', 'Tenure'], axis=1, inplace=True)
data_one_hot = pd.concat([data, gender_one_hot, country_one_hot, tenure_one_hot], axis=1)

Перекодировали столбцы

In [31]:
scaler = MinMaxScaler()
data_one_hot_scaled = data_one_hot.copy(deep=True)
data_one_hot_scaled[["Balance", "EstimatedSalary"]] = scaler.fit_transform(data_one_hot_scaled[["Balance", "EstimatedSalary"]])

Нормализовали данные

In [32]:
features = data.drop('Exited', axis=1)
target = data['Exited']

In [39]:
df_train, df_split = train_test_split(data, test_size=0.4, random_state = 12345)
df_valid, df_test = train_test_split(df_split, test_size=0.5, random_state = 12345)

In [40]:
features_train = df_train.drop(['Exited'], axis=1)
target_train = df_train['Exited']

In [41]:
features_test = df_test.drop(['Exited'], axis=1)
target_test = df_test['Exited']

Поделил выборки на 60/20/20, тренировочная, валидационная и тестовая соответсвенно.

In [43]:
logic_model = LogisticRegression(random_state=12345,class_weight='balanced')
logic_model.fit(features_train, target_train)
predictions_logic = logic_model.predict(features_valid)

print('Качество:', logic_model.score(features_valid, target_valid))
print('F1:', f1_score(predictions_logic, target_valid))
print('AUC-ROC:', roc_auc_score(target_valid, predictions_logic))

Качество: 0.652
F1: 0.4343302990897269
AUC-ROC: 0.6535588932119883


In [47]:
model_forest = RandomForestClassifier(random_state=12345)
model_forest.fit(features_train, target_train)
predictions_forest = model_forest.predict(features_valid)

print("Качество:", model_forest.score(features_valid, target_valid))
print("F1:", f1_score(predictions_forest, target_valid))
print('AUC-ROC:', roc_auc_score(target_valid, predictions_forest))

Качество: 0.9416
F1: 0.843010752688172
AUC-ROC: 0.8777859898028357


Случаный лес показывает лучшие метрики, будет прыгать от него.

## Шаг 3. Борьба с дисбалансом

In [70]:
def random_forest(features_train, features_valid, target_train, target_valid, cw):
    best_f1 = 0
    best_depth = 0
    best_est = 0
    best_roc_auc = 0
    for depth in range(8,20):
        for est in range(30,50,1):
            model = RandomForestClassifier(max_depth=depth, n_estimators=est, random_state=12345, class_weight=cw)
            model.fit(features_train, target_train)
            predictions = model.predict(features_valid)
            roc_auc = roc_auc_score(target_valid, predictions_forest)
            f1score = f1_score(predictions, target_valid)
            if f1score > best_f1:
                best_f1 = f1score
                best_depth = depth
                best_est = est
                best_roc_auc = roc_auc
    print('Глубина дерева:', best_depth, '\n', 'Количество деревьев:', best_est, '\n', '  F1 метрика:', best_f1, '\n', '  AUC-ROC метрика:', best_roc_auc)

In [71]:
random_forest(features_train, features_valid, target_train, target_valid, None)

Глубина дерева: 19 
 Количество деревьев: 43 
   F1 метрика: 0.843918191603875 
   AUC-ROC метрика: 0.8777859898028357


Нашли лучшие гаперпараметры модели.

In [54]:
def upsample(features, target, repeat):
    features_zeros = features[target == 0]
    features_ones = features[target == 1]
    target_zeros = target[target == 0]
    target_ones = target[target == 1]

    features_upsampled = pd.concat([features_zeros] + [features_ones] * repeat)
    target_upsampled = pd.concat([target_zeros] + [target_ones] * repeat)
    
    features_upsampled, target_upsampled = shuffle(
        features_upsampled, target_upsampled, random_state=12345)
    
    return features_upsampled, target_upsampled

функция для увеличения выборки.

In [55]:
def downsample(features, target, fraction):
    
    features_zeros = features[target == 0]
    features_ones = features[target == 1]
    target_zeros = target[target == 0]
    target_ones = target[target == 1]
    
    features_downsampled = pd.concat(
        [features_zeros.sample(frac=fraction, random_state=12345)] + [features_ones])
    target_downsampled = pd.concat(
        [target_zeros.sample(frac=fraction, random_state=12345)] + [target_ones])
    
    features_downsampled, target_downsampled = shuffle(
        features_downsampled, target_downsampled, random_state=12345)
    return features_downsampled, target_downsampled

Функция для уменьшения выборки.

In [57]:
features_train_upsampled, target_train_upsampled = upsample(features_train, target_train, 4)
features_train_balanced, target_train_balanced = downsample(features_train_upsampled, target_train_upsampled, 0.25)

In [58]:
data['Exited'].value_counts()

Exited
0    7963
1    2037
Name: count, dtype: int64

Соотношение 4 к 1

Проверим качество модели при уменьшении выборки, увеличении и балансировки классов.

In [59]:
random_forest(features_train_upsampled, features_valid, target_train_upsampled, target_valid, None)

Глубина дерева: 18     Количество деревьев: 37   F1 метрика: 0.8377281947261662   AUC-ROC метрика: 0.8777859898028357


In [60]:
random_forest(features_train_balanced, features_valid, target_train_balanced, target_valid, None)

Глубина дерева: 19     Количество деревьев: 30   F1 метрика: 0.6104152933421226   AUC-ROC метрика: 0.8777859898028357


In [61]:
random_forest(features_train, features_valid, target_train, target_valid, 'balanced')

Глубина дерева: 18     Количество деревьев: 44   F1 метрика: 0.8486140724946695   AUC-ROC метрика: 0.8777859898028357


Балансировка классов и увеличение выборки помогли достичь максимально высокой f1 метрики, где балансировка показала еще более лучший результат. Попробуем и увеличить выборку и сбалансировать классы, чтобы получить максимальную f1 метрику

In [62]:
random_forest(features_train_upsampled, features_valid, target_train_upsampled, target_valid, 'balanced')

Глубина дерева: 19     Количество деревьев: 43   F1 метрика: 0.8408163265306122   AUC-ROC метрика: 0.8777859898028357


Показатели стали хуже, но пусть будет, сейчас на тестовой выборке проеврим все три и выберем лучшую

## Шаг 4: Тест моделей

In [64]:
model = RandomForestClassifier(max_depth=18, n_estimators=44, random_state=12345, class_weight='balanced')
model.fit(features_train, target_train)
predictions = model.predict(features_test)
print(f1_score(predictions, target_test))
print('AUC-ROC:', roc_auc_score(target_test, predictions))

0.5022556390977443
AUC-ROC: 0.6736201993490948


In [65]:
model_upsample = RandomForestClassifier(max_depth=18, n_estimators=37, random_state=12345)
model_upsample.fit(features_train_upsampled, target_train_upsampled)
predictions = model_upsample.predict(features_test)
print(f1_score(predictions, target_test))
print('AUC-ROC:', roc_auc_score(target_test, predictions))

0.5584415584415584
AUC-ROC: 0.7122854988449505


In [67]:
model_upsample = RandomForestClassifier(max_depth=19, n_estimators=43, random_state=12345, class_weight='balanced')
model_upsample.fit(features_train_upsampled, target_train_upsampled)
predictions = model_upsample.predict(features_test)
print(f1_score(predictions, target_test))
print('AUC-ROC:', roc_auc_score(target_test, predictions))

0.5519053876478318
AUC-ROC: 0.7076435641783259


Так, на тестовой выборке не получилось добиться нужной f1 метрики, поэтомиу украду идею у Рената и скомбинирую данные.

In [72]:
features_combi = pd.concat([features_train, features_valid], axis=0)
target_combi = pd.concat([target_train, target_valid], axis=0)

In [73]:
model = RandomForestClassifier(max_depth=18, n_estimators=44, min_samples_leaf=4, random_state=12345, class_weight='balanced')
model.fit(features_combi, target_combi)
predictions = model.predict(features_test)
print(f1_score(predictions, target_test))
print('AUC-ROC:', roc_auc_score(target_test, predictions))

0.652014652014652
AUC-ROC: 0.7747023929986463


Вот теперь получилось добиться высокого результата.

## Вывод:

Загрузил данные, иследовал, удалил столбцы которые на не понадобятся, т.к. они не влияют на конечный результат.    
   
Проверил как сбалансированы данные в столбцах. Понял что есть несбалансированность.   
   
Далее разделил данные с учетом найденного дисбаланса в классах и нормализовал их.   
   
Начал создавать и обучать модели, решил проверить логистическую регерессию и случавйный лес, как ни странно лучший результат показал случайный лес с f1 метрикой равной 0.84, accuracy равной 0.94 и AUC-ROC метрикой равной 0.877  
   
Далее написал функцию для подбора лучших гиперпараметров для обучения модели и обучил модель получив требуемый результат с f1 метрикой равной 0.843.   
   
Попробовал уменьшить и увеличить выборку, при увеличении выборки и балансировке классов вместе с увеличением выборки получил нужный результат, но на тестовой выборке результат получился неудовлетворительным.
   
При тестировании модели не получил нужного результата и решил обучить модель на объедененных данных из обучающей и валидационных выборок. Получив удовлетворительную f1 метрику равную 0.65, также улучшив AUC-ROC метрику до значения 0.774 закончил выполнение задания.